File: server/src/routes/emails.js
import express from 'express';
import db from '../db/database.js';
import { requireAuth } from '../middleware/auth.js';

const router = express.Router();
router.use(requireAuth);

router.get('/', (req, res) => {
  const { label, unread } = req.query;
  let sql = 'SELECT * FROM emails WHERE user_id = ?';
  const params = [req.user.id];
  if (label) { sql += ' AND label = ?'; params.push(label); }
  if (unread === '1') sql += ' AND is_read = 0';
  sql += ' ORDER BY received_at DESC';
  const emails = db.prepare(sql).all(...params);
  res.json({ emails });
});

router.patch('/:id', (req, res) => {
  const existing = db.prepare('SELECT * FROM emails WHERE id = ? AND user_id = ?').get(req.params.id, req.user.id);
  if (!existing) return res.status(404).json({ error: 'Email not found' });

  const fields = ['is_read', 'is_starred', 'label'];
  const updates = [];
  const values = [];
  for (const f of fields) {
    if (f in req.body) {
      updates.push(`${f} = ?`);
      values.push(f === 'is_read' || f === 'is_starred' ? (req.body[f] ? 1 : 0) : req.body[f]);
    }
  }
  if (!updates.length) return res.json({ email: existing });
  values.push(req.params.id, req.user.id);
  db.prepare(`UPDATE emails SET ${updates.join(', ')} WHERE id = ? AND user_id = ?`).run(...values);
  const email = db.prepare('SELECT * FROM emails WHERE id = ?').get(req.params.id);
  res.json({ email });
});

router.delete('/:id', (req, res) => {
  db.prepare('DELETE FROM emails WHERE id = ? AND user_id = ?').run(req.params.id, req.user.id);
  res.json({ ok: true });
});

export default router;

File: server-py/app/api/emails.py
from fastapi import APIRouter, Depends, HTTPException, Request

from app.db import execute, query_all, query_one, row_to_dict
from app.deps import require_auth

router = APIRouter(prefix="/emails")


@router.get("")
def list_emails(request: Request, user=Depends(require_auth)):
    label = request.query_params.get("label")
    unread = request.query_params.get("unread")
    sql = "SELECT * FROM emails WHERE user_id = ?"
    params: list = [user["id"]]
    if label:
        sql += " AND label = ?"; params.append(label)
    if unread == "1":
        sql += " AND is_read = 0"
    sql += " ORDER BY received_at DESC"
    return {"emails": [row_to_dict(r) for r in query_all(sql, tuple(params))]}


_PATCH_FIELDS = ("is_read", "is_starred", "label")


@router.patch("/{email_id}")
def update_email(email_id: str, body: dict, user=Depends(require_auth)):
    existing = query_one("SELECT * FROM emails WHERE id = ? AND user_id = ?", (email_id, user["id"]))
    if not existing:
        raise HTTPException(404, "Email not found")
    updates: list[str] = []
    values: list = []
    for f in _PATCH_FIELDS:
        if f in body:
            updates.append(f"{f} = ?")
            v = body[f]
            if f in ("is_read", "is_starred"):
                v = 1 if v else 0
            values.append(v)
    if not updates:
        return {"email": row_to_dict(existing)}
    values.extend([email_id, user["id"]])
    execute(
        f"UPDATE emails SET {', '.join(updates)} WHERE id = ? AND user_id = ?",
        tuple(values),
    )
    return {"email": row_to_dict(query_one("SELECT * FROM emails WHERE id = ?", (email_id,)))}


@router.delete("/{email_id}")
def delete_email(email_id: str, user=Depends(require_auth)):
    execute("DELETE FROM emails WHERE id = ? AND user_id = ?", (email_id, user["id"]))
    return {"ok": True}
